src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp

```diff
@@ -345,86 +200,192 @@ uint G1ConcurrentRefine::max_num_threads() {
   return G1ConcRefinementThreads;
 }
 
-static size_t calc_new_green_zone(size_t green,
-                                  double logged_cards_scan_time,
-                                  size_t processed_logged_cards,
-                                  double goal_ms) {
-  // Adjust green zone based on whether we're meeting the time goal.
-  // Limit to max_green_zone.
-  const double inc_k = 1.1, dec_k = 0.9;
-  if (logged_cards_scan_time > goal_ms) {
-    if (green > 0) {
-      green = static_cast<size_t>(green * dec_k);
+void G1ConcurrentRefine::update_pending_cards_target(double logged_cards_time_ms,
```

Comment:
That it might leave the value unchanged when the data isn't good is an
implementation detail, not part of the API. A common case where this arises is
during application startup, when there might not be much or any GC refinement
to do (little or no oldgen).  The caller should just always treat it as doing
an update, and not do something different depending on a try_ result.